File: sht1/argonfluid/gmclj.c
/**********************************************************************
 *
 * File: gmclj.c
 *
 * Create random ("gas") initial configuration for NVT-Monte Carlo of
 * Lennard-Jonesium
 *
 * 12-May-2007 (MN)
 * 17-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpo;
  char fname[BSIZE];
  unsigned short dummy[3];
  unsigned short *seed;
  int n,nt,ntjob,ntprint,ntskip;
  int i;
  double disp,dr,rho,t;
  double c;
  double *x,*y,*z;

  /* User input */

  printf("              n=");
  getval_i(&n);
  printf("            rho=");
  getval_d(&rho);
  printf("              t=");
  getval_d(&t);
  printf("           disp=");
  getval_d(&disp);
  printf("             dr=");
  getval_d(&dr);
  printf("         ntskip=");
  getval_i(&ntskip);
  printf(" ntprint/ntskip=");
  getval_i(&ntprint);
  printf("   ntjob/ntskip=");
  getval_i(&ntjob);

  strcpy(fname,"mclj_in.dat");
  printf("          fname=[mclj_in.dat] ");
  getval_s(fname);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Random positions */

  c=cbrt(n/rho);
  for(i=0;i<=n-1;i++) {
    x[i]=c*(drand48()-0.5);
    y[i]=c*(drand48()-0.5);
    z[i]=c*(drand48()-0.5);
  }

  /* RNG seed */

  seed=seed48(dummy);

  /* Write startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&rho,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/

File: lj-canonical/fortran/gmclj.f90
!**********************************************************************!
!
! File: gmclj.f90
!
! Create random ("gas") initial configuration for NVT-Monte Carlo of 
! Lennard-Jonesium
!
! 13-Apr-1997 (MN)
! 17-Apr-2012
!
!**********************************************************************!

program gmclj

!**********************************************************************!

  implicit none

  integer,parameter::iout=2

  character(len=80)::fname
  integer::i,n,nran,nt,ntjob,ntprint,ntskip
  integer,dimension(:),allocatable::iran
  real::c,disp,dr,ran,rho,t
  real,dimension(:),allocatable::x,y,z

! User input

  write(unit=*,fmt="(a)",advance="no") "              n="
  read(unit=*,fmt=*) n
  write(unit=*,fmt="(a)",advance="no") "            rho="
  read(unit=*,fmt=*) rho
  write(unit=*,fmt="(a)",advance="no") "              t="
  read(unit=*,fmt=*) t
  write(unit=*,fmt="(a)",advance="no") "           disp="
  read(unit=*,fmt=*) disp
  write(unit=*,fmt="(a)",advance="no") "             dr="
  read(unit=*,fmt=*) dr
  write(unit=*,fmt="(a)",advance="no") "         ntskip="
  read(unit=*,fmt=*) ntskip
  write(unit=*,fmt="(a)",advance="no") " ntprint/ntskip="
  read(unit=*,fmt=*) ntprint
  write(unit=*,fmt="(a)",advance="no") "   ntjob/ntskip="
  read(unit=*,fmt=*) ntjob
  write(unit=*,fmt="(a)",advance="no") "          fname=[mclj_in.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="mclj_in.dat"
  end if

! Allocate arrays

  call random_seed(size=nran)

  allocate(iran(nran),x(n),y(n),z(n))

! Random positions

  c=(n/rho)**(1.0/3.0)
  do i=1,n
    call random_number(ran)
    x(i)=c*(ran-0.5)
    call random_number(ran)
    y(i)=c*(ran-0.5)
    call random_number(ran)
    z(i)=c*(ran-0.5)
  end do

! RNG seed

  call random_seed(get=iran)

! Write startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dr,rho,t
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program gmclj

!**********************************************************************!
